compiler/rustc_data_structures/src/vec_cache.rs

```diff
@@ -0,0 +1,266 @@
+        // * value is initialized since we saw a >= 2 index above.
+        // * `V: Copy`, so safe to read.
+        let value = unsafe { slot.byte_add(offset_of!(Slot<V>, value)).cast::<V>().read() };
+        Some((value, index))
+    }
+
+    /// Returns true if this successfully put into the map.
+    #[inline]
+    fn put<V>(&self, buckets: &[AtomicPtr<Slot<V>>; 21], value: V, extra: u32) -> bool {
+        static LOCK: std::sync::Mutex<()> = std::sync::Mutex::new(());
+
+        // SAFETY: `bucket_idx` is ilog2(u32).saturating_sub(11), which is at most 21, i.e.,
+        // in-bounds of buckets.
+        let bucket = unsafe { buckets.get_unchecked(self.bucket_idx) };
+        let mut ptr = bucket.load(Ordering::Acquire);
+        let _allocator_guard;
+        if ptr.is_null() {
+            // If we load null, then acquire the global lock; this path is quite cold, so it's cheap
+            // to use a global lock.
+            _allocator_guard = LOCK.lock();
+            // And re-load the value.
+            ptr = bucket.load(Ordering::Acquire);
+        }
+
+        // OK, now under the allocator lock, if we're still null then it's definitely us that will
+        // initialize this bucket.
+        if ptr.is_null() {
+            let bucket_layout =
+                std::alloc::Layout::array::<Slot<V>>(self.entries as usize).unwrap();
+            // SAFETY: Always >0 entries in each bucket.
+            let allocated = unsafe { std::alloc::alloc_zeroed(bucket_layout).cast::<Slot<V>>() };
+            if allocated.is_null() {
+                std::alloc::handle_alloc_error(bucket_layout);
+            }
+            bucket.store(allocated, Ordering::Release);
+            ptr = allocated;
+        }
+        assert!(!ptr.is_null());
+
+        // SAFETY: `index_in_bucket` is always in-bounds of the allocated array.
+        assert!(self.index_in_bucket < self.entries);
+        let slot = unsafe { ptr.add(self.index_in_bucket) };
+
+        let index_and_lock =
+            unsafe { &*slot.byte_add(offset_of!(Slot<V>, index_and_lock)).cast::<AtomicU32>() };
+        match index_and_lock.compare_exchange(0, 1, Ordering::AcqRel, Ordering::Acquire) {
+            Ok(_) => {
+                // We have acquired the initialization lock. It is our job to write `value` and
+                // then set the lock to the real index.
+
+                unsafe {
+                    slot.byte_add(offset_of!(Slot<V>, value)).cast::<V>().write(value);
+                }
+
+                index_and_lock.store(extra.checked_add(2).unwrap(), Ordering::Release);
+
+                true
+            }
+
+            // Treat "initializing" as actually initialized: we lost the race and should skip
+            // any updates to this slot. In practice this should be unreachable since we're guarded
+            // by an external lock that only allows one initialization for any given query result.
+            Err(1) => unreachable!(),
```

Comment:
we lock the *whole bucket* when first initializing it. Writing a yet empty slot of an existing bucket does not lock from what I can tell :thinking:

This should both mean that this is reachable unless "In practice this should be unreachable since we're guarded by an external lock that only allows one initialization for any given query result." refers to an implementation detail of *users* of `VecCache` and not of its API itself.

It should also mean that you could put the whole bucket initialization routine into a separate `#[cold]` function and only need to keep `LOOK` inside of that function, dropping the lock before writing to the slot here